### PR TITLE
[System] Import System.Diagnostics.Process

### PR DESCRIPTION
**Pull request by ludovic-henry**

### CODE REVIEW

**Comment by ludovic-henry** [comment]
@akoeplinger if you want to review, please do

File: mcs/class/System/ReferenceSources/EnvironmentHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+namespace System
+{
+	internal static class EnvironmentHelpers
+	{
+		internal static bool IsWindowsVistaOrAbove()
+		{
+			// Method should match the logic of the internal Environment.IsWindowsVista property in mscorlib
+			// If this method turns out to be heavily used we might want to cache at least part of this value.
+			OperatingSystem os = Environment.OSVersion;
+			return os.Platform == PlatformID.Win32NT && os.Version.Major >= 6;
```
Comment (akoeplinger) [inline mcs/class/System/ReferenceSources/EnvironmentHelpers.cs]: we can hardcode `true` here since Mono is Vista+ since some time now (and I incidentally have a bug assigned to make this clear on the website :smile: )

File: mono/io-layer/processes.c [hunk 2]
```diff
@@ -2754,7 +2754,32 @@ process_wait (gpointer handle, guint32 timeout, gboolean alertable)
...
-	g_assert (mp);
+	if (!mp) {
+		guint32 start, now;
+
+		/* We've already done this one */
+		MONO_TRACE (G_LOG_LEVEL_DEBUG, MONO_TRACE_IO_LAYER, "%s (%p, %u): waiting on non-child process", __func__, handle, timeout);
+
+		for (start = now = mono_msec_ticks (); timeout == INFINITE || now - start < timeout || now == start; now = mono_msec_ticks ()) {
+			int status;
+			pid_t res = waitpid (pid, &status, WNOHANG);
```
Comment (kumpera) [inline mono/io-layer/processes.c]: This doesn't looks particularly related to the change, it better go to a separate commit. 

I'm not sure this is a behavior we want to support, I rather throw an exception and force the user code
to do the polling.

Comment (ludovic-henry) [inline mono/io-layer/processes.c]: Without this, `HasExited` would be broken, due to the way it is implemented in ReferenceSource. If you revert that, the test `System.Diagnostics.ProcessTest.GetProcesses_StartTime` will crash with an assertion error

File: mono/metadata/process.h [hunk 3]
```diff
@@ -34,54 +34,40 @@ typedef struct
 typedef struct
 {
 	MonoObject object;
-	MonoString *arguments;
-	gpointer error_dialog_parent_handle;
 	MonoString *filename;
-	MonoString *verb;
+	MonoString *arguments;
```
Comment (kumpera) [inline mono/metadata/process.h]: Must bump corlib version

File: external/referencesource [hunk 4]
```diff
@@ -1 +1 @@
-Subproject commit b945d0b0a8a582e4d1ab581869b0c9215c7c3ca0
+Subproject commit b04e2857af0c5120ddd9785f701f89605e0dc3e7
```
Comment (kumpera) [inline external/referencesource]: Can you push your RS changes somewhere?

File: mcs/class/System/System.Diagnostics/Process.cs [hunk 5]
```diff
@@ -1253,65 +867,6 @@ public static Process Start(string fileName, string arguments, string username,
 		}
 #endif // MONO_FEATURE_PROCESS_START
 
-		public override string ToString()
-		{
-			return(base.ToString() + " (" + this.ProcessName + ")");
```
Comment (akoeplinger) [inline mcs/class/System/System.Diagnostics/Process.cs]: The implementation in referencesource only does base.ToString() now.

File: mcs/class/System/System.Diagnostics/Process.cs [hunk 6]
```diff
@@ -447,55 +288,13 @@ public Process ()
 			}
 		}
 
-		[MonoLimitation ("Under Unix, only root is allowed to raise the priority.")]
-		[DesignerSerializationVisibility (DesignerSerializationVisibility.Hidden)]
-		[MonitoringDescription ("The relative process priority.")]
-		public ProcessPriorityClass PriorityClass {
```
Comment (akoeplinger) [inline mcs/class/System/System.Diagnostics/Process.cs]: The setter in referencesource does some unnecessary checks for Win2k that can be removed.

File: external/referencesource (hunk 4 above)
Comment (akoeplinger) [inline external/referencesource]: https://github.com/mono/referencesource/commit/b04e2857af0c5120ddd9785f701f89605e0dc3e7

File: mono/io-layer/processes.c (hunk 2 above)
Comment (kumpera) [inline mono/io-layer/processes.c]: That brings the question of how much value bringing Process has. Since it's 100% implemented on top of the windows behavior.

I'm looking at the code for HasExited and it polls for the exit status by using a zero timeout.

I think we can support just that and avoid polling on ppl's back.

Comment (ludovic-henry) [inline mono/io-layer/processes.c]: By importing it from referencesource, we should get rid of some race conditions in our implementation of Process. XS particularly suffers from it.

File: mono/metadata/process.h (hunk 3 above)
Comment (ludovic-henry) [inline mono/metadata/process.h]: done

**Comment by ludovic-henry** [comment]
build

File: mcs/class/System/System.Diagnostics/Process.cs (hunk 5 above)
Comment (ludovic-henry) [inline mcs/class/System/System.Diagnostics/Process.cs]: this is fixed by the way we now treat `FEATURE_PAL` in referencesource `Process.cs

File: mcs/class/System/System.Diagnostics/Process.cs (hunk 6 above)
Comment (ludovic-henry) [inline mcs/class/System/System.Diagnostics/Process.cs]: done

File: mono/io-layer/processes.c (hunk 2 above)
Comment (ludovic-henry) [inline mono/io-layer/processes.c]: Since this code path is also only used by `HasExited`, we will never poll more than once. If you prefer, we can assert that timeout is 0, it will at least fix that case, and not introduce any other chance of continuously polling.

**Comment by ludovic-henry** [comment]
@kumpera I added some stress tests as requested.

File: mono/tests/process-stress-1.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
+
+using System;
```
Comment (akoeplinger) [inline mono/tests/process-stress-1.cs]: these should have the license headers

File: mono/tests/Makefile.am [hunk 8]
```diff
@@ -1414,6 +1414,14 @@ test-console-output: console-output.exe
 	@diff -w console-output.exe.stdout $(srcdir)/console-output.exe.stdout.expected \
 		&& diff -w console-output.exe.stderr $(srcdir)/console-output.exe.stderr.expected
 
+PROCESS_STRESS_TESTS=	\
+		process-stress-1.exe	\
+		process-stress-2.exe	\
+		process-stress-3.exe
+
+test-process-stress: $(PROCESS_STRESS_TESTS) test-runner.exe
```
Comment (akoeplinger) [inline mono/tests/Makefile.am]: are they intended to run on-demand or should we hook them up so that they run during make check?

Comment (ludovic-henry) [inline mono/tests/Makefile.am]: They are more intended to run on demand when working on the Process implementation to check an array of common race conditions.

File: mono/io-layer/processes.c (hunk 2 above)
Comment (kumpera) [inline mono/io-layer/processes.c]: Sounds good, let's do that.

**Comment by kumpera** [comment]
LGTM, removing the polling can be done on a follow up commit.

**Comment by akoeplinger** [comment]
process-stress-2.exe and process-stress-3.exe fail with exit code 5 on OSX for me.

**Comment by ludovic-henry** [comment]
@kumpera I removed the polling, it will now assert that the timeout is 0, and simply check if the process has exited.

@akoeplinger that should now be fixed, it was a race condition in the test code

**Comment by akoeplinger** [comment]
@ludovic-henry yep, passes now on OSX and Ubuntu :)

File: mono/tests/Makefile.am (hunk 8 above)
Comment (akoeplinger) [inline mono/tests/Makefile.am]: I'd prefer having them run in CI, those on-demand things tend to be forgotten and we don't catch regressions.

**Comment by ludovic-henry** [comment]
build

**Comment by ludovic-henry** [comment]
@monojenkins merge

**Comment by ludovic-henry** [comment]
@kumpera should we backport that to 4.3.2?

**Comment by kumpera** [comment]
I think it's too late for that. This is a major change.
